pkg/controller/trainingruntime_controller.go

```diff
@@ -0,0 +1,149 @@
+	}
+	log := ctrl.LoggerFrom(ctx).WithValues("trainingRuntime", klog.KObj(&runtime))
+	ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling TrainingRuntime")
+
+	trainJobs := &trainer.TrainJobList{}
+	if err := r.client.List(ctx, trainJobs, client.InNamespace(runtime.Namespace), client.MatchingFields{
+		idxer.TrainJobRuntimeRefKey: runtime.Name,
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+	if !ctrlutil.ContainsFinalizer(&runtime, constants.ResourceInUseFinalizer) && len(trainJobs.Items) != 0 {
+		ctrlutil.AddFinalizer(&runtime, constants.ResourceInUseFinalizer)
+		return ctrl.Result{}, r.client.Update(ctx, &runtime)
+	} else if !runtime.DeletionTimestamp.IsZero() && len(trainJobs.Items) == 0 {
+		ctrlutil.RemoveFinalizer(&runtime, constants.ResourceInUseFinalizer)
+		return ctrl.Result{}, r.client.Update(ctx, &runtime)
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *TrainingRuntimeReconciler) NotifyTrainJobUpdate(oldJob, newJob *trainer.TrainJob) {
+	var runtimeNSName *types.NamespacedName
+	switch {
+	case oldJob != nil && newJob != nil:
+		// UPDATE Event.
+		return
+	case oldJob == nil:
+		// CREATE Event.
+		if trainjob.RuntimeRefIsTrainingRuntime(newJob.Spec.RuntimeRef) {
+			runtimeNSName = &types.NamespacedName{Namespace: newJob.Namespace, Name: newJob.Spec.RuntimeRef.Name}
+		}
+	default:
+		// DELETE Event.
+		if trainjob.RuntimeRefIsTrainingRuntime(oldJob.Spec.RuntimeRef) {
+			runtimeNSName = &types.NamespacedName{Namespace: oldJob.Namespace, Name: oldJob.Spec.RuntimeRef.Name}
+		}
+	}
+	if runtimeNSName != nil {
+		r.nonRuntimeObjectUpdateCh <- event.TypedGenericEvent[iter.Seq[types.NamespacedName]]{
+			Object: slices.Values([]types.NamespacedName{*runtimeNSName}),
+		}
+	}
+}
+
+type nonRuntimeObjectHandler struct{}
+
+var _ handler.TypedEventHandler[iter.Seq[types.NamespacedName], reconcile.Request] = (*nonRuntimeObjectHandler)(nil)
+
+func (h *nonRuntimeObjectHandler) Create(context.Context, event.TypedCreateEvent[iter.Seq[types.NamespacedName]], workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+func (h *nonRuntimeObjectHandler) Update(context.Context, event.TypedUpdateEvent[iter.Seq[types.NamespacedName]], workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+func (h *nonRuntimeObjectHandler) Delete(context.Context, event.TypedDeleteEvent[iter.Seq[types.NamespacedName]], workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+}
+func (h *nonRuntimeObjectHandler) Generic(_ context.Context, e event.TypedGenericEvent[iter.Seq[types.NamespacedName]], q workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+	for nsName := range e.Object {
+		q.AddAfter(reconcile.Request{NamespacedName: nsName}, time.Second)
+	}
+}
+
+func (r *TrainingRuntimeReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	return builder.TypedControllerManagedBy[reconcile.Request](mgr).
+		Named("trainingruntime_controller").
+		WithOptions(options).
+		WatchesRawSource(source.TypedKind(
```

Comment:
Could `For` be used here?

Reply:
Unfortunately, the TypedManager does not provide `For`.